File: backend/apps/dashboards/dashboards.py
import json
import os
from contextlib import asynccontextmanager
from datetime import datetime
from uuid import uuid4
from pydantic import BaseModel

from backend.config.Apps import SubApp
from backend.core.db.PydanticStore import PydanticStore
from backend.core.shared_structs.dashboard.Dashboard import Dashboard
from backend.core.shared_structs.dashboard.DashboardLayout import DashboardLayout
from backend.apps.agents.agents import get_all_sessions, delete_session
from backend.apps.settings.settings import load_settings
from backend.apps.dashboards.generate_dashboard_name import generate_dashboard_name
from backend.ports import NINE_ROUTER_PORT
from swarm_debug import debug
from typing import Optional
from backend.config.paths import DB_ROOT


DASHBOARD_STORE: PydanticStore[Dashboard] = PydanticStore[Dashboard](
    model_cls=Dashboard,
    data_dir=os.path.join(DB_ROOT, "dashboards"),
    id_field="id",
    dump_mode="json",
    not_found_detail="Dashboard not found",
)

@asynccontextmanager
async def dashboards_lifespan():
    yield


dashboards = SubApp("dashboards", dashboards_lifespan)


@dashboards.router.get("/list")
async def list_dashboards():
    all_dashboards = DASHBOARD_STORE.load_all()
    all_dashboards.sort(key=lambda d: d.updated_at or d.created_at, reverse=True)
    items = []
    for d in all_dashboards:
        dumped = d.model_dump(mode="json")
        items.append({
            "id": dumped["id"],
            "name": dumped.get("name", "Untitled"),
            "auto_named": dumped.get("auto_named", False),
            "created_at": dumped.get("created_at"),
            "updated_at": dumped.get("updated_at"),
            "thumbnail": dumped.get("thumbnail"),
        })
    return {"dashboards": items}


class DashboardCreate(BaseModel):
    name: str = "Untitled Dashboard"

@dashboards.router.post("/create")
async def create_dashboard(body: DashboardCreate):
    dashboard = Dashboard(name=body.name)
    DASHBOARD_STORE.save(dashboard)
    return dashboard.model_dump(mode="json")


# TODO: Maybe parse the output of get_all_sessions into actual Agent objects?
@dashboards.router.post("/{dashboard_id}/generate-name")
async def generate_name(dashboard_id: str):
    dashboard = DASHBOARD_STORE.load(dashboard_id)

    if not dashboard.auto_named and dashboard.name != "Untitled Dashboard":
        return {"name": dashboard.name, "auto_named": dashboard.auto_named}

    sessions_resp = await get_all_sessions(dashboard_id=dashboard_id)
    sessions = sessions_resp.get("SESSIONS", [])

    prompts = []
    for session in sessions:
        for msg in session.get("messages", {}).get("messages", []):
            if msg.get("role") == "user" and isinstance(msg.get("content"), str) and msg["content"].strip():
                prompts.append(msg["content"].strip()[:200])
                break

    if not prompts:
        return {"name": dashboard.name, "auto_named": dashboard.auto_named}

    settings = load_settings()
    fallback = prompts[0][:40]
    try:
        generated = await generate_dashboard_name(
            prompts=prompts,
            api_key=settings.anthropic_api_key,
            nine_router_port=NINE_ROUTER_PORT,
        )
        if generated:
            fallback = generated
    except Exception as e:
        debug(f"[dashboards.generate_name] ERROR: Dashboard name generation failed, using fallback: {e}")

    dashboard.name = fallback
    dashboard.auto_named = True
    dashboard.updated_at = datetime.now()
    DASHBOARD_STORE.save(dashboard)
    return {"name": dashboard.name, "auto_named": True}


@dashboards.router.get("/{dashboard_id}")
async def get_dashboard(dashboard_id: str):
    dashboard = DASHBOARD_STORE.load(dashboard_id)
    return dashboard.model_dump(mode="json")




class DashboardUpdate(BaseModel):
    name: Optional[str] = None
    layout: Optional[DashboardLayout] = None
    thumbnail: Optional[str] = None

@dashboards.router.put("/{dashboard_id}")
async def update_dashboard(dashboard_id: str, body: DashboardUpdate):
    dashboard = DASHBOARD_STORE.load(dashboard_id)
    if body.name is not None:
        dashboard.name = body.name
        dashboard.auto_named = False
    if body.layout is not None:
        dashboard.layout = body.layout
    if body.thumbnail is not None:
        dashboard.thumbnail = body.thumbnail
    dashboard.updated_at = datetime.now()
    DASHBOARD_STORE.save(dashboard)
    return dashboard.model_dump(mode="json")


@dashboards.router.delete("/{dashboard_id}")
async def delete_dashboard(dashboard_id: str):
    DASHBOARD_STORE.load(dashboard_id)  # confirm it exists (raises 404 if not)

    sessions_resp = await get_all_sessions(dashboard_id=dashboard_id)
    for session in sessions_resp.get("SESSIONS", []):
        try:
            await delete_session(session["session_id"])
        except Exception:
            debug(f"[dashboards.delete_dashboard] ERROR: Failed to delete session {session.get('session_id')} during dashboard deletion")

    DASHBOARD_STORE.delete(dashboard_id)
    return {"ok": True}


@dashboards.router.post("/{dashboard_id}/duplicate")
async def duplicate_dashboard(dashboard_id: str):
    source = DASHBOARD_STORE.load(dashboard_id)
    source_data = source.model_dump(mode="json")
    new_id = uuid4().hex
    now = datetime.now().isoformat()

    new_dashboard = {
        **source_data,
        "id": new_id,
        "name": f"{source_data.get('name', 'Untitled')} (copy)",
        "created_at": now,
        "updated_at": now,
        "layout": {
            "cards": {},
            "view_cards": source_data.get("layout", {}).get("view_cards", {}),
            "browser_cards": source_data.get("layout", {}).get("browser_cards", {}),
        },
    }
    with open(os.path.join(DB_ROOT, "dashboards", f"{new_id}.json"), "w") as f:
        json.dump(new_dashboard, f, indent=2)

    return new_dashboard

File: frontend/src/shared/backend-bridge/apps/dashboards.ts
import { createAsyncThunk } from '@reduxjs/toolkit';
import { API_BASE } from '@/shared/backend-bridge/base_routes';

const DASHBOARDS_API: string = `${API_BASE}/dashboards`;

export interface DashboardLayout {
  cards?: Record<string, unknown>;
  view_cards?: Record<string, unknown>;
  browser_cards?: Record<string, unknown>;
  expanded_session_ids?: string[];
}

export interface Dashboard {
  id: string;
  name: string;
  auto_named: boolean;
  created_at: string;
  updated_at: string;
  thumbnail?: string | null;
  layout?: DashboardLayout;
}

// ---------------------------------------------------------------------------
// Dashboard CRUD
// ---------------------------------------------------------------------------


const list_dashboards_endpoint: string = `${DASHBOARDS_API}/list`;
async function list_dashboards_function(): Promise<Dashboard[]> {
  const res = await fetch(list_dashboards_endpoint, {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data.dashboards as Dashboard[];
}
export const LIST_DASHBOARDS = createAsyncThunk(
  list_dashboards_endpoint,
  list_dashboards_function,
);


const create_dashboard_endpoint: string = `${DASHBOARDS_API}/create`;
async function create_dashboard_function(name: string = 'Untitled Dashboard'): Promise<Dashboard> {
  const res = await fetch(create_dashboard_endpoint, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  });
  const data = await res.json();
  return data as Dashboard;
}
export const CREATE_DASHBOARD = createAsyncThunk(
  create_dashboard_endpoint,
  create_dashboard_function,
);


const generate_dashboard_name_endpoint: string = `${DASHBOARDS_API}/generate-name`;
async function generate_dashboard_name_function(dashboardId: string): Promise<{ id: string; name: string; auto_named: boolean }> {
  const res = await fetch(`${DASHBOARDS_API}/${dashboardId}/generate-name`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return { id: dashboardId, name: data.name as string, auto_named: data.auto_named as boolean };
}
export const GENERATE_DASHBOARD_NAME = createAsyncThunk(
  generate_dashboard_name_endpoint,
  generate_dashboard_name_function,
);


const get_dashboard_endpoint: string = `${DASHBOARDS_API}/get`;
async function get_dashboard_function(dashboardId: string): Promise<Dashboard> {
  const res = await fetch(`${DASHBOARDS_API}/${dashboardId}`, {
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data as Dashboard;
}
export const GET_DASHBOARD = createAsyncThunk(
  get_dashboard_endpoint,
  get_dashboard_function,
);


const update_dashboard_endpoint: string = `${DASHBOARDS_API}/update`;
async function update_dashboard_function(args: {
  dashboardId: string;
  name?: string;
  layout?: Record<string, unknown>;
  thumbnail?: string;
}): Promise<Dashboard> {
  const { dashboardId, ...updates } = args;
  const res = await fetch(`${DASHBOARDS_API}/${dashboardId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(updates),
  });
  if (!res.ok) throw new Error(`Dashboard update failed: ${res.status}`);
  const data = await res.json();
  return data as Dashboard;
}
export const UPDATE_DASHBOARD = createAsyncThunk(
  update_dashboard_endpoint,
  update_dashboard_function,
);


const delete_dashboard_endpoint: string = `${DASHBOARDS_API}/delete`;
async function delete_dashboard_function(dashboardId: string): Promise<string> {
  const res = await fetch(`${DASHBOARDS_API}/${dashboardId}`, {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
  });
  await res.json();
  return dashboardId;
}
export const DELETE_DASHBOARD = createAsyncThunk(
  delete_dashboard_endpoint,
  delete_dashboard_function,
);


const duplicate_dashboard_endpoint: string = `${DASHBOARDS_API}/duplicate`;
async function duplicate_dashboard_function(dashboardId: string): Promise<Dashboard> {
  const res = await fetch(`${DASHBOARDS_API}/${dashboardId}/duplicate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
  });
  const data = await res.json();
  return data as Dashboard;
}
export const DUPLICATE_DASHBOARD = createAsyncThunk(
  duplicate_dashboard_endpoint,
  duplicate_dashboard_function,
);
